### PR TITLE
[replicate] support simpler device_id

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99090
* #100216
* #100131

Allow passing in `device_id=[device]` regardless of CPU or GPU. We
modify the kwarg as needed to pass to DDP.

Differential Revision: [D44979763](https://our.internmc.facebook.com/intern/diff/D44979763/)

### CODE REVIEW

File: torch/distributed/_composable/replicate.py [hunk 1]
```diff
@@ -85,8 +86,21 @@ def init_helper(self) -> None:
         self._collect_params(self.module)
         # Only saved for testing
         replicate.state(self.module)._names = self._names
+        if "device_ids" in self.kwargs:
```
Comment (awgu) [inline torch/distributed/_composable/replicate.py:89]: Should we even preserve the argument to be called `device_ids`? That is a relic from single process multiple device days right?

Can we just call it `device_id` and share the same type as for FSDP?

File: torch/distributed/_composable/replicate.py [hunk 2]
```diff
@@ -85,8 +86,21 @@ def init_helper(self) -> None:
...
+            # replicate() supports a small usability enhancement where
+            # device_ids=[self.device] can also be passed in for CPU so users
```
Comment (awgu) [inline torch/distributed/_composable/replicate.py:91]: nit: IIUC, this comment has some implied context on what is `self.device`. What is `self` here? The module?

File: torch/distributed/_composable/replicate.py [hunk 3]
```diff
@@ -85,8 +86,21 @@ def init_helper(self) -> None:
...
+            # don't have to code change for CPU / GPU runs.
+            device_ids = self.kwargs["device_ids"]
+            if device_ids is not None and not isinstance(device_ids, list):
+                raise RuntimeError(
+                    f"Expected device_ids arg to be Optional[List[Union[int, torch.device]]], got {device_ids}"
+                )
+            device_id = device_ids[0]
+            if isinstance(device_id, torch.device) and device_id.type == "cpu":
```
Comment (awgu) [inline torch/distributed/_composable/replicate.py:99]: So basically, if the user passes in the CPU device, then we treat it as a no-op?

File: torch/distributed/_composable/replicate.py [hunk 4]
```diff
@@ -85,8 +86,21 @@ def init_helper(self) -> None:
...
         # Only saved for testing
         replicate.state(self.module)._names = self._names
+        if "device_ids" in self.kwargs:
+            # replicate() supports a small usability enhancement where
+            # device_ids=[self.device] can also be passed in for CPU so users
+            # don't have to code change for CPU / GPU runs.
+            device_ids = self.kwargs["device_ids"]
+            if device_ids is not None and not isinstance(device_ids, list):
+                raise RuntimeError(
+                    f"Expected device_ids arg to be Optional[List[Union[int, torch.device]]], got {device_ids}"
```
Comment (awgu) [inline torch/distributed/_composable/replicate.py:96]: nit: It might be more useful of a message to say something like
"Expected device_ids args to be None or a list of devices or device indices but got {device_ids}"
but up to you which you think is clearer.

File: torch/distributed/_composable/replicate.py [hunk 5]
```diff
@@ -85,8 +86,21 @@ def init_helper(self) -> None:
...
+                )
+            device_id = device_ids[0]
+            if isinstance(device_id, torch.device) and device_id.type == "cpu":
+                self.kwargs["device_ids"] = None
 
         self._ddp = DistributedDataParallel(self._param_list, **self.kwargs)
+        replicate.state(self.module)._ddp_weakref = weakref.ref(self._ddp)
```
Comment (awgu) [inline torch/distributed/_composable/replicate.py:103]: Could you explain why this change is added? Is this only to test the `device_ids` change? Do we need this if only for testing?

Comment (rohan-varma) [reply torch/distributed/_composable/replicate.py:103]: yes, only for testing

File: torch/distributed/_composable/replicate.py (hunk 1 above)
Comment (rohan-varma) [reply torch/distributed/_composable/replicate.py:89]: Sure, but device_id won't really mean the same thing in replicate and FSDP then. FSDP does more things with device_id such as move the module to that device if its on CPU. replicate() might do such things in the future, but not at the moment. At the current moment, DDP device_id arg is mostly used only for moving inputs to the right device.

Comment (awgu) [reply torch/distributed/_composable/replicate.py:89]: I think that difference between `replicate` and `fully_shard` is fine. The frontend arg name being the same might be more important than unifying semantics right now. What do you think?

I was mainly thinking that we do not have to preserve the legacy `device_ids` (plural) if we are defining a new API now.

File: torch/distributed/_composable/replicate.py (hunk 5 above)
Comment (awgu) [reply torch/distributed/_composable/replicate.py:103]: Could we just test for the expected behavior then? I am okay with adding this weakref, but I worry if it is only for this particular test, then we may just have this unused attribute that becomes forgotten in the future. If you think it will be useful in the future too, then I am down to keep it.

File: torch/distributed/_composable/replicate.py (hunk 1 above)
Comment (rohan-varma) [reply torch/distributed/_composable/replicate.py:89]: Sure, will switch this to device_id

File: torch/distributed/_composable/replicate.py (hunk 5 above)
Comment (rohan-varma) [reply torch/distributed/_composable/replicate.py:103]: I think it will be useful more generally because there could be future changes that require us to validate some setting on the DDP instance (e.g. communication hook). It's also quite hard to test whether DDP device_ids was populated correctly, as at this point device_ids is really only used for moving inputs to the right device. I guess we could have a test that ensures this, but having a weakref and checking the attribute to be as expected seems better to me.

Comment (awgu) [reply torch/distributed/_composable/replicate.py:103]: Sounds good to me! Maybe we add a comment before defining the `_ddp_weakref` mentioning that this can be useful for testing if you think that can help clarify.

File: torch/distributed/_composable/replicate.py [hunk 6]
```diff
@@ -25,6 +26,9 @@ def replicate(
         >>> replicate(module)
     """
     torch._C._log_api_usage_once("torch.distributed.replicate")
+    if "device_id" in kwargs:
+        if not isinstance(kwargs["device_id"], (int, torch.device)):
+            raise RuntimeError(f"Expected device_id to be int or torch.device, but got {type(kwargs['device_id'])}")
```
Comment (zhaojuanmao) [inline torch/distributed/_composable/replicate.py:31]: nit: seems to duplicate with below assertion

**Review by zhaojuanmao — APPROVED** [review]
